### DISCUSSION TITLE
Adding item to UnsafeList<T> doesn't increase Count?

### DISCUSSION THREAD

**Discussion by ClemGG** [genaray/Arch#225] [Q&A]
EDIT: Sorry for the code formatting, Github seems to refuse to format it correctly.

Hello, I'm working on an InputManager for a custom engine for my game. Due to the lack of an UnsafeDirectory<T> type, I've been using UnsafeLists to store the ID of an Action and the handles containing the references to the Actions themselves.

In the following example, calling AddAction() doesn't seem to properly register the key to the IDs list. Using a breakpoint in Visual Studio reveals the item is added to the underlying UnsafeArray<>, but the actual Count isn't increased. This causes a problem in the GetButtonEvent() method where the Contains() method always returns false as its Count stays at 0.

What's strange about this is that creating and UnsafeList<> directly inside the AddAction() method, and adding the key to it successfully increases its Count property; but doing it with one of the UnsafeList<> fields doesn't work.

I'm at a loss about what to do. Is it an oversight on my part? Any suggestion would be welcome.

`
using System;
using Arch.LowLevel;
using Retard.Core.Models.ValueTypes;
using Retard.Core.ViewModels.Input;

namespace Retard.Engine.Models.Assets.Input
{
    internal readonly struct InputHandles
    {
        private readonly UnsafeList<NativeString> _buttonStateHandlesIDs;
        private readonly UnsafeList<InputActionButtonStateHandles> _buttonStateHandles;

        public InputHandles()
        {
            this._buttonStateHandlesIDs = new UnsafeList<NativeString>(1);
            this._buttonStateHandles = new UnsafeList<InputActionButtonStateHandles>(1);
        }

        public void AddButtonEventIfNotExists(NativeString key)
        {
            if (!this._buttonStateHandlesIDs.Contains(key))
            {
                this._buttonStateHandlesIDs.Add(key);   // Count stays at 0 ?

                var started = InputManager.Instance.ActionButtonResources.Add(delegate
                { });
                var performed = InputManager.Instance.ActionButtonResources.Add(delegate
                { });
                var finished = InputManager.Instance.ActionButtonResources.Add(delegate
                { });

                this._buttonStateHandles.Add(new InputActionButtonStateHandles(started, performed, finished));
            }
        }
       
        public ref readonly InputActionButtonStateHandles GetButtonEvent(NativeString key)
        {
            // Always returns false since Count stays at 0
            if (!this._buttonStateHandlesIDs.Contains(key))
            {
                throw new ArgumentException($"Error : The key doesn't exist.");
            }

            return ref this._buttonStateHandles[this._buttonStateHandlesIDs.IndexOf(key)];
        }
    }
}
`

**Comment by genaray** [comment]
Is it possible that Its because the field is readonly? Remove it and try again :)

**Reply by ClemGG** [reply]
Hi @genaray, thank you for your time. I've already tried that, on both the struct and the fields, but nothing changed; and I would have probably had a compilation error if that was the issue. Just to be sure, I tried to replace the NativeString struct by a simple value type like char, just in case the underlying values couldn't be copied for some reason, but that didn't work either.
